packages/liveblocks-core/src/connection.ts

```diff
@@ -767,18 +771,31 @@ function createConnectionStateMachine<T extends BaseAuthResult>(
 
   const sendHeartbeat: Target<Context, Event | BuiltinEvent, State> = {
     target: "@ok.awaiting-pong",
-    effect: (ctx: Context) => {
+    effect: (ctx) => {
       ctx.socket?.send("ping");
     },
   };
 
+  const maybeHeartbeat: Target<Context, Event | BuiltinEvent, State> = () => {
+    // If the browser tab isn't visible currently, ask the application if going
+    // zombie is fine
+    const doc = typeof document !== "undefined" ? document : undefined;
+    const canZombie =
+      doc?.visibilityState === "hidden" && delegates.canZombie();
+    return canZombie ? "@idle.zombie" : sendHeartbeat;
+  };
+
   machine
-    .addTimedTransition("@ok.connected", HEARTBEAT_INTERVAL, sendHeartbeat)
+    .addTimedTransition("@ok.connected", HEARTBEAT_INTERVAL, maybeHeartbeat)
     .addTransitions("@ok.connected", {
-      NAVIGATOR_OFFLINE: sendHeartbeat, // Don't take the browser's word for it when it says it's offline. Do a ping/pong to make sure.
+      NAVIGATOR_OFFLINE: maybeHeartbeat, // Don't take the browser's word for it when it says it's offline. Do a ping/pong to make sure.
       WINDOW_GOT_FOCUS: sendHeartbeat,
     });
 
+  machine.addTransitions("@idle.zombie", {
+    WINDOW_GOT_FOCUS: "@connecting.backoff", // When in zombie state, the client will try to wake up automatically when the window regains focus
```

Comment:
I can't seem to figure out how does this get triggered?

Reply:
oh, I see, WINDOW_GOT_FOCUS already existed, not part of this PR. I answer my own questions 😂